### PR TITLE
`NormalizesTo`: return nested goals to caller

### PR DESCRIPTION
**Pull request by lcnr**
Fixes the regression of `paperclip-core`. see https://hackmd.io/IsVAafiOTAaPIFcUxRJufw for more details.

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs [hunk 1]
```diff
@@ -93,13 +95,26 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
...
+        // When normalizing, we've replaced the expected term with an unconstrained
+        // inference variable. This means that we dropped information which could
+        // have been important. We handle this by instead returning the nested goals
+        // to the caller, where they are then handled.
+        //
+        // As we return all ambiguous nested goals, we can ignore the certainty returned
+        // by `try_evaluate_added_goals()`.
+        let (certainty, normalization_nested_goals) = if self.is_normalizes_to_goal {
+            let NestedGoals { normalizes_to_goals, goals } = std::mem::take(&mut self.nested_goals);
+            assert!(normalizes_to_goals.is_empty());
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs:107]: Can we assert that if `goal_certainty` is ambiguous, `goals.len() > 0`? Seems like a good way to make sure that "we can ignore the certainty returned by `try_evaluate_added_goals()`" remains sound.

File: compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs [hunk 2]
```diff
@@ -319,6 +331,20 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
...
     ) -> Result<(bool, Certainty), NoSolution> {
+        let (normalization_nested_goals, has_changed, certainty) =
+            self.evaluate_goal_raw(goal_evaluation_kind, source, goal)?;
+        assert!(normalization_nested_goals.is_empty());
+        Ok((has_changed, certainty))
+    }
+
+    /// FIXME(-Znext-solver=coinduction): `_source` is currently unused but will
+    /// be necessary once we implement the new coinduction approach.
+    fn evaluate_goal_raw(
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs:342]: Needs a comment explaining wtf this does lol

File: compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs [hunk 3]
```diff
@@ -319,6 +331,20 @@ impl<'a, 'tcx> EvalCtxt<'a, 'tcx> {
...
         goal: Goal<'tcx, ty::Predicate<'tcx>>,
     ) -> Result<(bool, Certainty), NoSolution> {
+        let (normalization_nested_goals, has_changed, certainty) =
+            self.evaluate_goal_raw(goal_evaluation_kind, source, goal)?;
+        assert!(normalization_nested_goals.is_empty());
+        Ok((has_changed, certainty))
+    }
+
+    /// FIXME(-Znext-solver=coinduction): `_source` is currently unused but will
+    /// be necessary once we implement the new coinduction approach.
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt/mod.rs:341]: don't need to be doc, this needs real doc comments tho

```suggestion
    // FIXME(-Znext-solver=coinduction): `_source` is currently unused but will
    // be necessary once we implement the new coinduction approach.
```

File: compiler/rustc_type_ir/src/predicate_kind.rs [hunk 4]
```diff
@@ -148,19 +148,20 @@ pub enum PredicateKind<I: Interner> {
     /// Used for coherence to mark opaque types as possibly equal to each other but ambiguous.
     Ambiguous,
 
-    /// The alias normalizes to `term`. Unlike `Projection`, this always fails if the alias
-    /// cannot be normalized in the current context.
+    /// This should only used inside of the mew solver for `AliasRelate` and expects
```
[the remark this answers is not in the document]
Comment (lcnr) [reply compiler/rustc_type_ir/src/predicate_kind.rs:151]: MEW!

**Comment by BoxyUwU** [comment]
if we never evaluate goals inside of `normalizes-to` are we able to make normalizes-to a tcx query since we don't actually care about whatever goals are in the search graph for evaluating it?

File: compiler/rustc_type_ir/src/predicate_kind.rs [hunk 5]
```diff
@@ -148,19 +148,20 @@ pub enum PredicateKind<I: Interner> {
     /// Used for coherence to mark opaque types as possibly equal to each other but ambiguous.
     Ambiguous,
 
-    /// The alias normalizes to `term`. Unlike `Projection`, this always fails if the alias
-    /// cannot be normalized in the current context.
+    /// This should only used inside of the new solver for `AliasRelate` and expects
```
Comment (lqd) [inline compiler/rustc_type_ir/src/predicate_kind.rs:151]:
```suggestion
    /// This should only be used inside of the new solver for `AliasRelate` and expects
```

File: compiler/rustc_type_ir/src/predicate_kind.rs [hunk 6]
```diff
@@ -148,19 +148,20 @@ pub enum PredicateKind<I: Interner> {
...
+    /// the `term` to be an unconstrained inference variable.
     ///
-    /// `Projection(<T as Trait>::Assoc, ?x)` results in `?x == <T as Trait>::Assoc` while
-    /// `NormalizesTo(<T as Trait>::Assoc, ?x)` results in `NoSolution`.
-    ///
-    /// Only used in the new solver.
+    /// The alias normalizes to `term`. Unlike `Projection`, this always ails if the
```
Comment (lqd) [inline compiler/rustc_type_ir/src/predicate_kind.rs:154]:
```suggestion
    /// The alias normalizes to `term`. Unlike `Projection`, this always fails if the
```

**Review by lqd — COMMENTED** [review]
High-value review comments™

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**